pkg/storaged/dialogx.jsx

```diff
@@ -529,10 +529,38 @@ class SizeSliderElement extends React.Component {
     }
 
     render() {
-        let { val, max, onChange } = this.props;
+        let { val, max, round, onChange } = this.props;
         let { unit } = this.state;
 
-        const change_slider = (f) => onChange(Math.round(f * max));
+        const round_size = (value) => {
+            if (round) {
+                if (typeof round == "function")
+                    value = round(value);
+                else
+                    value = Math.round(value / round) * round;
+            }
+
+            // As a special case, if the user types something that
+            // looks like the maximum (or minimum) when formatted,
+            // always use exactly the maximum (or minimum).  Otherwise
+            // we have the confusing possibility that with the exact
+            // same string in the text input, the size is sometimes
+            // too large (or too small) and sometimes not.
+
+            function sanitize(limit) {
+                var fmt = cockpit.format_number(limit / unit);
+                var parse = +fmt * unit;
+
+                if (value == parse)
+                    value = limit;
+            }
+
+            sanitize(max);
+
+            return Math.round(value);
+        };
```

Comment:
I don't understand why this always calls `Math.round()` in the end. If a `round` property function is given, isn't that supposed to already "properly" round?  If it's given as a number, it would also already be `Math.round`'ed, by the inner `else`. So shouldn't this only be done on `if (!round)`?

Reply:
> I don't understand why this always calls Math.round() in the end. 

Good question.  I think this happened because of the way the code evolved, by adding more features to a basic Math.round.  There is no harm, though, and this way, we have a clear guarantee that the number is always an integer.  (If it isn't, there will be obscure D-Bus error messages later.)